### DISCUSSION TITLE
Support API key generation

### DISCUSSION THREAD

**Discussion by FelixZY** [supabase/supabase#12328] [Feature Requests]
# Feature request

Please support generating user-tied API keys via `supabase-js`/`GoTrue`

## Is your feature request related to a problem? Please describe.

I'm developing an open API. Users can sign into a web portal and generate API tokens to be used when communicating with the API. Users can generate multiple tokens, e.g. if they want to use different keys for Android and iOS client applications and track their usage separately.

Currently, I can find no good way of generating such tokens/API keys without exposing secrets or breaking RLS.

## Describe the solution you'd like

It should be possible to create an "authorized token"/"api key"/other static string for signed in users. Such tokens could be stored in an `access_tokens(user_id uuid, access_token varchar, created_at timezonetz)` table in the `auth` schema. This token should then be possible to supply in the request header instead of a jwt when accessing supabase. Queries using this token should be subject to the same RLS policies as if the user's jwt token was sent. Additionally, it should be possible to discern if the request was made using an API key and in that case, which one. This could be done e.g. by introducing `auth.api_key()` and would allow me to create additional tables and RLS policies to limit the access a given API key provides (e.g. "this token can only be used to read the user's public git repositories") myself.

## Describe alternatives you've considered

1. Using my service key to violate RLS on a custom, protected, `api_keys(user uuid, key varchar)` table (still cannot impersonate a user + RLS is now useless),
2. minting my own JWTs (hard, error prone and feels like asking for trouble), or
3. Limiting access to generic access only (using anon key, and forcing me to keep this key more secret).

## Additional context

#4419

**Comment by FelixZY** [comment]
**EDIT**: I rewrote this entire solution (as noted further down (https://github.com/supabase/supabase/issues/7186#issuecomment-1214179734)). Check out https://gist.github.com/FelixZY/0aef530690458b381b8100afa19202c8 instead.

---

Ok, so I found a workable workaround:

1. Create function `gen_api_token`
```pgsql
CREATE
OR REPLACE FUNCTION gen_api_token() RETURNS text LANGUAGE plpgsql AS $$
BEGIN
-- Replace $NUMBER_OF_BYTES with what fits your needs. I would reccommend 128 bits or more (i.e 16+)
RETURN (SELECT encode(gen_random_bytes($NUMBER_OF_BYTES), 'base64') AS token);
END;
$$;
```
2. Create table `apitokens(id bigint, token varchar, user uuid, created_at timestamptz)`. Make `token`' column unique with default value `gen_api_token()`.
3. Create function `http_header`
```pgsql
CREATE
OR REPLACE FUNCTION http_header(name text) RETURNS text LANGUAGE sql stable AS $$
SELECT NULLIF(current_setting('request.headers', true)::json->>name, '')::text;
$$;
```
4. Add read policy for `apitokens`:
```pgsql
BEGIN;
  ALTER POLICY "Allow tokens to see themselves for verification" ON "public"."apitokens" USING (http_header('x-api-token') = token);
COMMIT;
```
5. Add RLS policies as required:
```
-- Untested. I've wrapped the SELECT call into an `api_user()` function for ease of use
user_id = uid() OR (SELECT user_id FROM apitokens WHERE apitokens.token = http_header('x-api-token')) = user_id
```

*You can now make API requests using your anon key + `"x-api-token`"-header with your custom api token and RLS!*

I still think this usecase could be nice to support out-of-the-box, especially to use standardized DOS protections etc.

**Comment by dotlouis** [comment]
Thank you @FelixZY for the help and suggestions I've been looking to do something similar for a bit and now have successfully implemented it 🙌.

I agree, it took a lot of work for a common use case. and it would be great to have it supported by default.

**Comment by FelixZY** [comment]
I'm working on a substantial improvement to my above solution (I'll update this issue with it when it is done). In the meanwhile, I would like to draw attention to https://github.com/supabase/supabase/discussions/6380 and https://github.com/supabase/supabase/discussions/8175 as these questions are related to the administration of API keys.

**Comment by FelixZY** [comment]
So after having some more time to work on my implementation, I found it lacking in many ways. You'd need to provide an additional header, could not use `auth.uid()`, were recognized as anonymous, etc. For this reason I went back to the drawing board and dove deeper into the documentation. The result is a substantially improved implementation which I have posted as a gist here: https://gist.github.com/FelixZY/0aef530690458b381b8100afa19202c8 . Please let me know what you think, if you have any questions, etc.

**Comment by ausmurp** [comment]
@FelixZY thank you for taking the time to make that script. Have you thought about creating a PR for the admins to review and potentially put into the auth schema as you suggested originally?

**Comment by FelixZY** [comment]
@ausmurp I have, but I don't think my suggestion fits properly enough for a PR. My main (but not only) issue is the workaround to get logging working, which really should use the cloudflare edge logs instead (but there is no way to do that atm AFAIK).

I would rather share my solution here as a temporary workaround for others and discuss my findings with supabase's maintainers to work out something that would support, rather than clash with, the product.

I'm not opposed to rewriting the script to a working PR, I'm just not sure Supabase is capable enough for such a rewrite to be possible yet.

**Comment by j4w8n** [comment]
@FelixZY @dotlouis @ausmurp 

I created a set of functions to do this. You probably have your solutions set up by now, but I wanted to share and get your feedback.

https://gist.github.com/j4w8n/25d233194877f69c1cbf211de729afb2

**Comment by david-r-cox** [comment]
[We're](https://github.com/integrated-reasoning) launching [KeyHippo](https://keyhippo.com/) specifically to solve this so that you can write RLS policies like:

```SQL
CREATE POLICY "allow_owner_access"
ON "public"."table_name"
USING (
  (SELECT "auth"."uid"() = "table_name"."owner_id")
  OR (SELECT "auth"."key_uid"() = "table_name"."owner_id")
);
```

KeyHippo aside, I'm happy to help anyone who's trying to add API keys to their app.

**Reply by ekremkenter** [reply]
Cool feature but I wish it was a contribution to OSS instead of a new product. Wish you the best though!

**Reply by jumploops** [reply]
@ekremkenter I believe it is OSS, MIT licensed: https://github.com/integrated-reasoning/KeyHippo
